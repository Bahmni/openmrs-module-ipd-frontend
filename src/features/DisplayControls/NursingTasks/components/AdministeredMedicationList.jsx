import React from "react";
import {
  StructuredListWrapper,
  StructuredListBody,
  StructuredListRow,
  StructuredListCell,
} from "carbon-components-react";
import PropTypes from "prop-types";
import { getMedicationDetails } from "../utils/TaskTileUtils";
import ClockIcon from "../../../../icons/clock.svg";
import Administered from "../../../../icons/administered.svg";
import NotAdministered from "../../../../icons/not-administered.svg";
import "../styles/AdministeredMedicationList.scss";

const AdministeredMedicationList = ({ list }) => {
  const getAdministeredTime = (date) => {
    const hours = date.getHours().toString().padStart(2, "0");
    const minutes = date.getMinutes().toString().padStart(2, "0");
    return `${hours}:${minutes}`;
  };

  const getMedicationStatus = (status) => {
    return status === "completed" ? "Done" : "Skipped";
  };

  const getStatusIcon = (status) => {
    return status === "completed" ? <Administered /> : <NotAdministered />;
  };

  return (
    <StructuredListWrapper className="administered-list-body">
      <StructuredListBody>
        {Object.keys(list).map((key, index) => {
          return (
            <StructuredListRow key={index}>
              <StructuredListCell className="drug-name-cell">
                <div>{list[key].displayName}</div>
                <span>{getMedicationDetails(list[key])}</span>
              </StructuredListCell>
              <StructuredListCell className="administered-time-cell">
                <ClockIcon className="clock-icon" />
                <span className="time">
                  {getAdministeredTime(new Date(list[key].actualTime))}
                </span>
              </StructuredListCell>
              <StructuredListCell>
                {getStatusIcon(list[key].status)}
                <span className="status" data-testid="status-name-cell">
                  {getMedicationStatus(list[key].status)}
                </span>
              </StructuredListCell>
            </StructuredListRow>
          );
        })}
      </StructuredListBody>
    </StructuredListWrapper>
  );
};

AdministeredMedicationList.propTypes = {
  list: PropTypes.object.isRequired,
};
export default AdministeredMedicationList;
