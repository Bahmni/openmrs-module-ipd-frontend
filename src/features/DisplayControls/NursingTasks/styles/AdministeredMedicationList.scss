.bx--structured-list {
  margin: 0;
}

.administered-list-body {
  .time,
  .status {
    position: relative;
    padding-left: 8px;
  }

  .time {
    top: -3px;
  }

  .status {
    top: -4px;
  }

  .drug-name-cell {
    width: 55%;
  }

  .administered-time-cell {
    --day-border: 0px !important;
  }
}
