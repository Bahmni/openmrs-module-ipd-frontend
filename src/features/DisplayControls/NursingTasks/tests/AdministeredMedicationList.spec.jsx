import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import AdministeredMedicationList from "../components/AdministeredMedicationList";

const mockAdministeredMedicationsList = {
  "8040ab44-2100-4e4e-91ad-50a5715050da": {
    displayName: "Liquid Paraffin",
    doseType: "Drop",
    dosage: 2,
    route: "Topical",
    startTime: "16:00",
    isSelected: true,
    actualTime: "2023-12-15T12:30:21.731Z",
    status: "completed",
  },
};

describe("AdministeredMedicationList", () => {
  it("should render AdministeredMedicationList", () => {
    const { container } = render(
      <AdministeredMedicationList list={mockAdministeredMedicationsList} />
    );
    expect(container).toMatchSnapshot();
  });

  test("should call getStatus function", () => {
    const mockGetStatus = jest.fn();
    render(
      <AdministeredMedicationList list={mockAdministeredMedicationsList} />
    );
    screen.debug();
    mockGetStatus("completed");
    expect(mockGetStatus).toHaveBeenCalledTimes(1);
    expect(screen.getByText("Done")).toBeInTheDocument();
  });
});
