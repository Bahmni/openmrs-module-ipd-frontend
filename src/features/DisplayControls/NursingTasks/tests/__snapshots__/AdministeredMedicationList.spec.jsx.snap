// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdministeredMedicationList should render AdministeredMedicationList 1`] = `
<div>
  <div
    aria-label="Structured list section"
    class="bx--structured-list administered-list-body"
    role="table"
  >
    <div
      class="bx--structured-list-tbody"
      role="rowgroup"
    >
      <div
        class="bx--structured-list-row"
        role="row"
      >
        <div
          class="drug-name-cell bx--structured-list-td"
          role="cell"
        >
          <div>
            Liquid Paraffin
          </div>
          <span>
            <div>
              <span>
                2
              </span>
              <span>
                 - 
                Drop
              </span>
              <span>
                 - 
                Topical
              </span>
            </div>
          </span>
        </div>
        <div
          class="administered-time-cell bx--structured-list-td"
          role="cell"
        >
          <test-file-stub
            classname="clock-icon"
          />
          <span
            class="time"
          >
            12:30
          </span>
        </div>
        <div
          class="bx--structured-list-td"
          role="cell"
        >
          <test-file-stub />
          <span
            class="status"
            data-testid="status-name-cell"
          >
            Done
          </span>
        </div>
      </div>
      <div
        class="bx--structured-list-row"
        role="row"
      >
        <div
          class="drug-name-cell bx--structured-list-td"
          role="cell"
        >
          <div>
            Adenosine 3 mg/mL Ampoule/Vial
          </div>
          <span>
            <div>
              <span>
                5mg
              </span>
              <span>
                 - 
                Intravenous
              </span>
            </div>
          </span>
        </div>
        <div
          class="administered-time-cell bx--structured-list-td"
          role="cell"
        >
          <test-file-stub
            classname="clock-icon"
          />
          <span
            class="time"
          >
            14:00
          </span>
        </div>
        <div
          class="bx--structured-list-td"
          role="cell"
        >
          <test-file-stub />
          <span
            class="status"
            data-testid="status-name-cell"
          >
            Skipped
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
